packages/cardpay-sdk/sdk/reward-manager/base.ts

```diff
@@ -480,25 +477,65 @@ export default class RewardManager {
 The owner of reward safe ${safeAddress} is ${rewardSafeOwner}, but the signer is ${from}`
       );
     }
-    let weiAmount = new BN(amount);
-    if (weiAmount.gt(safeBalance)) {
-      throw new Error(
-        `Insufficient funds inside reward safe. safeBalance = ${fromWei(safeBalance)} and amount = ${fromWei(amount)}`
+
+    let estimate;
+    let withdrawPayload;
+    let weiAmount;
+    if (amount) {
+      //when amount is given, we assume that amount is string in wei that has already had gasCost deducted from it
+      let weiAmount = new BN(amount);
+      if (weiAmount.gt(safeBalance)) {
+        throw new Error(
+          `Insufficient funds inside reward safe. safeBalance = ${fromWei(safeBalance)} and amount = ${fromWei(amount)}`
+        );
+      }
+      let withdraw = await rewardSafeDelegate.methods.withdraw(rewardManagerAddress, tokenAddress, to, weiAmount);
+      withdrawPayload = withdraw.encodeABI();
+      estimate = await gasEstimate(
+        this.layer2Web3,
+        safeAddress,
+        rewardSafeDelegateAddress,
+        '0',
+        withdrawPayload,
+        Operation.DELEGATECALL,
+        tokenAddress
+      );
+      let gasCost = new BN(estimate.safeTxGas).add(new BN(estimate.baseGas)).mul(new BN(estimate.gasPrice));
+      if (safeBalance.lt(gasCost.add(weiAmount))) {
+        throw new Error(
+          `Reward safe does not have enough to pay for gas when withdrawing rewards. The reward safe ${safeAddress} balance for token ${tokenAddress} is ${fromWei(
+            safeBalance
+          )}, amount being withdrawn is ${fromWei(amount)}, the gas cost is ${fromWei(gasCost)}`
+        );
+      }
+    } else {
+      //when amount is NOT given, we use safeBalance - gasCost as the withdraw amount
+      //Note: gasCost is estimated with safeBalance not the actual withdraw amount
+      weiAmount = safeBalance;
+      let withdraw = await rewardSafeDelegate.methods.withdraw(rewardManagerAddress, tokenAddress, to, weiAmount);
```

Comment:
@habdelra Note this part where the `gasCost` is a function of a `gasEstimation` that uses the **safeBalance** safeBalance. I am not sure whether this will affect us. Typically, before we did `gasEstimation` that uses the **safeBalance - gasCost`. Is it a point to do the gas estimation again before executing the transaction?

Reply:
well the thing is that the relay server uses the on chain oracle to determine how much native coin the gas token is worth. in the time between when you signed the txn and then executed the txn, the value of the gas token may have updated in teh oracle making the amount being deducted for gas insufficient. for stable coin this would almost never happen, but for volatile coins like CARD this is a possibility.

Reply:
(this is actually the reason that we introduced the rate lock idea for how prepaid cards work--this same vulnerability exists when trying to empty out a safe and still have enough left over to pay for gas when using a gas token that can fluctuate dramatically in value)